File: server-mh/src/Match/Handler/MatchPhases/ActionPhase.ts
import { PhaseBase } from "./PhaseBase";
import { MatchContext } from "../Models/MatchContex";
import { ACTIONSELECT_BOT_TIMEOUT_SECONDS,MATCH_TICK_RATE,ACTIONSELECT_HUMAN_TIMEOUT_SECONDS } from "../Consts";
import { ClientOpCode, Phase } from "../Enums";
export class ActionPhase extends PhaseBase {

    public override Start(context: MatchContext): void {
        const currentPlayer =
            context.state.players[context.state.turnState.currentPlayer];

        context.state.diceState.waitingForActionSelect = true;

        context.state.tickCounter = currentPlayer.playerState.isBot
            ? ACTIONSELECT_BOT_TIMEOUT_SECONDS * MATCH_TICK_RATE
            : ACTIONSELECT_HUMAN_TIMEOUT_SECONDS * MATCH_TICK_RATE;
    }

    public override Update(context: MatchContext): void {

       
        context.state.tickCounter--;

        if (context.state.tickCounter <= 0) {
            this.SelectRandomAction(context);
            return;
        }

        if (this.HandleSelectAction(context)) {
            return;
        }
    }

  

    private HandleSelectAction(context: MatchContext): boolean {

        for (const message of context.messages) {

            if (message.opCode !== ClientOpCode.SelectAction)
                continue;

            const currentPlayer =
                context.state.players[context.state.turnState.currentPlayer];

            if (message.sender.userId !== currentPlayer.userId)
                return false;

            const index = Number(context.nk.binaryToString(message.data));

            if (
                !Number.isInteger(index) ||
                index < 0 ||
                index >= context.state.availableActions!.length
            ) {
                return false;
            }

            context.state.selectedAction = index;
            context.state.pendingPhase = Phase.Resolution;

            return true;
        }

        return false;
    }

    private SelectRandomAction(context: MatchContext): void {

        const actions = context.state.availableActions!;

        if (actions.length === 0) {
            throw new Error("Invariant violation: availableActions is empty.");
        }

        context.state.selectedAction =
            Math.floor(Math.random() * actions.length);

        context.state.pendingPhase = Phase.Resolution;
    }
}

File: Client-unity/Ludo/Assets/Scripts/OfflineMatchHandler/ActionPhase.cs
using UnityEngine;

namespace Ludo.Offline
{
    public class ActionPhase : PhaseBase
    {
        private bool _actionSelected;
        private int _selectedActionIndex;
        public ActionPhase()
        {
            GameManager.Instance.GamePlayHandler.GamePlayEvents.ActionSelected += OnActionSelected;
        }
        public void OnActionSelected(int actionIndex)
        {
            _selectedActionIndex = actionIndex;
            _actionSelected = true;
        }

        public override void Start(MatchContext context)
        {
            var state = context.State;

            var currentPlayer =
                state.Players[(int)state.TurnState.CurrentPlayer.Value];

            state.DiceState.WaitingForActionSelect = true;

            state.TickCounter = currentPlayer.PlayerState.IsBot
                ? MatchConstants.ActionSelectBotTimeoutSeconds *
                  MatchConstants.MatchTickRate
                : MatchConstants.ActionSelectHumanTimeoutSeconds *
                  MatchConstants.MatchTickRate;

            _actionSelected = false;
        }

        public override void Update(MatchContext context)
        {
            context.State.TickCounter--;

            if (context.State.TickCounter <= 0)
            {
                SelectRandomAction(context);
                return;
            }

            if (HandleSelectAction(context))
            {
                return;
            }
        }

        private bool HandleSelectAction(MatchContext context)
        {
            if (!_actionSelected)
                return false;

            _actionSelected = false;

            var state = context.State;

            if (state.AvailableActions == null ||
                state.AvailableActions.Count == 0)
            {
                return false;
            }

            if (_selectedActionIndex < 0 ||
                _selectedActionIndex >= state.AvailableActions.Count)
            {
                return false;
            }

            state.SelectedAction = _selectedActionIndex;

            state.DiceState.WaitingForActionSelect = false;

            state.PendingPhase = Phase.Resolution;

            return true;
        }

        private void SelectRandomAction(MatchContext context)
        {
            var state = context.State;

            if (state.AvailableActions == null ||
                state.AvailableActions.Count == 0)
            {
                throw new System.InvalidOperationException(
                    "Invariant violation: availableActions is empty."
                );
            }

            state.SelectedAction =
                Random.Range(0, state.AvailableActions.Count);

            state.DiceState.WaitingForActionSelect = false;

            state.PendingPhase = Phase.Resolution;
        }
    }
}